pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,615 @@
+	}
+
+	err = future.WaitForCompletion(context.TODO(), vmClient.Client)
+	if err != nil {
+		return nil, fmt.Errorf("waiting for operation returned: %s", err.Error())
+	}
+
+	vm, err := future.Result(vmClient)
+	if err != nil {
+		return nil, fmt.Errorf("decoding result: %s", err.Error())
+	}
+
+	// get the actual VM object filled in with additional data
+	vm, err = vmClient.Get(context.TODO(), config.ResourceGroup, machine.Spec.Name, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve updated data for VM %q: %s", machine.Spec.Name, err)
+	}
+
+	ipAddresses, err := getVMIPAddresses(context.TODO(), config, &vm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve IP addresses for VM %q: %s", machine.Spec.Name, err.Error())
+	}
+
+	status, err := getVMStatus(context.TODO(), config, machine.Spec.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve status for VM %q: %s", machine.Spec.Name, err.Error())
+	}
+
+	return &azureVM{vm: &vm, ipAddresses: ipAddresses, status: status}, nil
+}
+
+func (p *provider) Delete(machine *v1alpha1.Machine, instance instance.Instance) error {
+	config, _, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	// TODO: Disassociate and remove the resources in reverse order, to prevent
+	// orphaned NICs, IPs and disks in case the operation is interrupted in the middle.
+	glog.Infof("deleting VM %q", machine.Name)
+	if err = deleteVMsByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove public IP addresses of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting disks of VM %q", machine.Name)
+	if err = deleteDisksByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove disks of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting network interfaces of VM %q", machine.Name)
+	if err = deleteInterfacesByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove network interfaces of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting public IP addresses of VM %q", machine.Name)
+	if err = deleteIPAddressesByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove public IP addresses of machine %q: %s", machine.Name, err)
+	}
+
+	return nil
+}
+
+func getVMByUID(ctx context.Context, c *config, uid types.UID) (*compute.VirtualMachine, error) {
+	vmClient, err := getVMClient(c)
+	if err != nil {
+		return nil, err
+	}
+
+	list, err := vmClient.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var allServers []compute.VirtualMachine
+
+	for {
+		allServers = append(allServers, list.Values()...)
+		if list.NotDone() {
```

Comment:
move ` list.NotDone()` into the condition for the for loop

Reply:
fixed